### ISSUE TITLE
question

### ISSUE THREAD

**Issue by dimiy** [tranphuminhbkhn/MiLu#1]
sorry can you give your email address or where is it convenient for you to chat? I wanted to ask something about the framework and skills.

**Comment by tranphuminhbkhn** [comment]
Hi, my email is: tranphuminhkhn@gmail.com

**Comment by dimiy** [comment]
I get a message that your address was not found so I will ask a question here I hope that you do not mind
can you tell me resources where it is shown writing procts in deep learning from scratch (without frameworks) and where neural entities write from scratch?
Express your opinion on the given resources. you are welcome.
http://blog.ezyang.com/2019/05/pytorch-internals/
https://medium.com/@whitesell.ben/how-to-build-your-own-deep-learning-framework-its-easy-i-ll-prove-it-c859cb790386
https://end-to-end-machine-learning.teachable.com/p/write-a-neural-network-framework/
https://www.xinjianl.com/blog/tag/pytensor/page/4/
http://dlsys.cs.washington.edu/
https://course.fast.ai/videos/?lesson=8
is it possible to implement object recognition from scratch (with its own framework)? If so, what kind of project architecture do you need?
many thanks if you can help

**Comment by tranphuminhbkhn** [comment]
Hi,
Sorry, I mistakenly wrote. My email is: tranphuminhbkhn@gmail.com.

And about your question. I don't write neural from scratch, but tensor. You can see src/tensor.cpp, where i define tensor class. Because it's a course projects, a simple way i just define 2d tensor.

Tensor class include:
Properties
  - n, m: shape
  - d: value
  - grad: gradient
  - left: left child of tensor in graph
  - righ: right child of tensor in graph
  - isvar: is it variable? is it updated after each epoch?

Methods
  - some constructor.
  - run() compute tensor's value from descendants node of it. It like Tensor.run() in Tensorflow.
  - backward() compute tensor's gradient. It like Tensor.backward() in Pytorch.
  - *step: update all descendants variable tensor of it. Like Optimizer.step() in Pytorch.

And of course, it is possible to implement object recognition from scratch. But my framwork not, I don't write convolutional layer, because it's quite complicated.

Let me know if you have any questions.
